### ISSUE TITLE
ip() calls should get cached

### ISSUE THREAD

**Issue by TheRealFalcon** [canonical/pycloudlib#316]
Every Instance has an ip() method. In cloud-init, when tearing down, we have the option of keeping instance so that it can be further debugged. To make this easy, we print the IP of the instance being torn down. However, this can cause issues on "clouds" like LXD that dynamically fetch the IP. In my case, I took a snapshot of the instance first, so the instance was shutdown (but not removed). The call to get the IP then retries 150 times, eventually failing.

It's fine (not ideal) if the call fails, but waiting 150 tries is a bit much.

**Comment by holmanb** [comment]
The issue in cloud-init mentioned here was worked around in https://github.com/canonical/cloud-init/pull/4419, but a better failure path would still be nice. That said, IPs can (and do) change throughout the lifecycle of an instance, so I'm not sure that caching an IP really makes the most sense, because these values may become stale.

**Comment by aciba90** [comment]
I agree. A client can call `Instance.{add,remove}_network_interface` or modify IPs by other means, invalidating the cached one.

**Comment by holmanb** [comment]
Currently this issue is characterized as "pycloudlib might fail while getting ip dynamically"

The proposed solution is "cache the ip".

However, I think the issue is really just localized to the lxd [`LXDInstance.ip()` method](https://github.com/canonical/pycloudlib/blob/86c02b80ca0ab1a2e7230370db3244efbf2ce60b/pycloudlib/lxd/instance.py#L141).

Perhaps it would be better to recharacterize this issue as "`LXDInstance.ip()` misbehaves when an IP isn't available".

As for a solution, I don't have one to propose at this time, but I'm suspicious about why retrying parsing the IP would fix anything if the first attempt failed, and I don't think that raising a timeout error is the best behavior in the failure case - I'd suggest just return `None` or `""` and document this in the base class.

**Comment by TheRealFalcon** [comment]
> Perhaps it would be better to recharacterize this issue as "LXDInstance.ip() misbehaves when an IP isn't available".

+1. Cache the IP was a bad idea.

> but I'm suspicious about why retrying parsing the IP would fix anything if the first attempt failed

When starting the instance, not all clouds have an IP assigned the moment an instance is launched. In cloud-init, one of the first things we do after launching an instance is attempt to acquire an IP. If it isn't available yet, we need to retry. I'm open to moving that retry logic, but it would need to exist somewhere.

> and I don't think that raising a timeout error is the best behavior in the failure case - I'd suggest just return None or "" and document this in the base class.

I know that  pycloudlib is still very inconsistent in this approach, but I'd really like to standardize on raising exceptions anytime an operation fails or can't return an expected result. It allows us to encode a reason for the failure which is often needed depending on who is using the library. Even in this case, the reason we're not getting an IP back is going to be different on LXD vs say ec2, and that distinction may be important.

> As for a solution, I don't have one to propose at this time

Could better exceptions be part of the solution? For LXD, we could have a NoIpAndWithThisConfigurationYoullNeverGetOne exception vs other clouds may have a TheIpIsntAvailableRightNow exception (with better names of course) and cloud-init (or whoever) can act accordingly.

**Comment by aciba90** [comment]
Agreed, caching sounds not very practical.

> Could better exceptions be part of the solution? For LXD, we could have a NoIpAndWithThisConfigurationYoullNeverGetOne exception vs other clouds may have a TheIpIsntAvailableRightNow exception (with better names of course) and cloud-init (or whoever) can act accordingly.

I think exceptions with different meanings is the way to go in the long term.

> we have the option of keeping instance so that it can be further debugged. To make this easy, we print the IP of the instance being torn down ...  The call to get the IP then retries 150 times, eventually failing.

For this issue, thinking out loud, could we have a `get_ip_fail_fast` (with a better name for sure)? Such method would try to get the ip and if not present fail fast.
